dali/pipeline/operators/color/hsv.h

```diff
@@ -0,0 +1,281 @@
+      kernel_manager_.Resize(num_threads_, batch_size_);
+    }
+  }
+
+  ~Hsv() override = default;
+
+  DISABLE_COPY_MOVE_ASSIGN(Hsv);
+
+ protected:
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+
+  USE_OPERATOR_MEMBERS();
+  hsv::argument_t<Backend> hue_, saturation_, value_;
+  DALIDataType output_type_;
+  kernels::KernelManager kernel_manager_;
+};
+
+
+
+class HsvCpu : public Hsv<CPUBackend> {
+ public:
+  explicit HsvCpu(const OpSpec &spec) : Hsv(spec) {}
+
+  /*
+   * So that compiler wouldn't complain, that
+   * "overloaded virtual function `dali::Operator<dali::CPUBackend>::RunImpl`
+   * is only partially overridden in class `dali::HsvCpu`"
+   */
+  using Operator<CPUBackend>::RunImpl;
+
+  ~HsvCpu() override = default;
+
+
+  DISABLE_COPY_MOVE_ASSIGN(HsvCpu);
+
+
+ protected:
+  bool SetupImpl(std::vector<::dali::OutputDesc> &output_desc,
+                 const workspace_t<CPUBackend> &ws) override;
+
+
+  void RunImpl(Workspace<CPUBackend> &ws) override;
+
+
+ private:
+  mat3 transformation_matrix(float hue, float saturation, float value) const {
+    using namespace hsv;  // NOLINT
+    return Rgb2Yiq * compose_hue(hue) * compose_saturation(saturation) * compose_value(value) *
+           Yiq2Rgb;
+  }
+
+
+  template <typename Kernel, typename InputType>
+  kernels::TensorListShape<> CallSetup(const TensorVector<CPUBackend> &input, int instance_idx) {
+    kernels::KernelContext ctx;
+    kernels::TensorListShape<> sh = input.shape();
+    std::vector<kernels::TensorShape<>> shapes;
+    shapes.resize(sh.num_samples());
+    for (size_t i = 0; i < shapes.size(); i++) {
+      const auto tvin = view<const InputType, 3>(input[i]);
+      const auto tmat = transformation_matrix(hue_, saturation_, value_);
+      const auto reqs = kernel_manager_.Setup<Kernel>(instance_idx, ctx, tvin, tmat);
+      kernels::TensorListShape<> out_sh = reqs.output_shapes[0];
+      shapes[i] = out_sh.tensor_shape(0);
+    }
+    return {shapes};
+  }
+};
+
+
+
+class HsvGpu : public Hsv<GPUBackend> {
+ public:
+  explicit HsvGpu(const OpSpec &spec) : Hsv(spec) {}
+
+
+  ~HsvGpu() override = default;
+
+
+  DISABLE_COPY_MOVE_ASSIGN(HsvGpu);
+
+
+ protected:
+  bool SetupImpl(std::vector<::dali::OutputDesc> &output_desc,
+                 const workspace_t<GPUBackend> &ws) override;
+
+
+  void RunImpl(Workspace<GPUBackend> &ws) override;
+
+
+ private:
+  /**
+   * @brief Creates transformation matrices based on given args
+   */
+  std::vector<mat3>
+  determine_transformation(const std::vector<float> &hue, const std::vector<float> &saturation,
+                           const std::vector<float> &value) const {
+    using namespace hsv;  // NOLINT
+    std::vector<mat3> ret;
+    DALI_ENFORCE(hue.size() == saturation.size());
+    DALI_ENFORCE(saturation.size() == value.size());
+    for (size_t i = 0; i < hue.size(); i++) {
+      ret.emplace_back(Rgb2Yiq * compose_hue(hue[i]) * compose_saturation(saturation[i]) *
```

Comment:
1. The output has fixed, known size - you can resize it before filling and save multiple reallocations when growing the vector.
2. The names `compose_xxx` are not that helpful - just long.
3. All three transforms could be done in one go:
```
mat3 hsv_matrix_Yiq(float h_angle, float s_mul, float v_mul) {
  const float deg2rad = (M_PI / 180);
  float a = h_angle * deg2rad;
  float c = cos(a) * v_mul * s_mul;
  float s = sin(a) * v_mul * s_mul;
  return {{
    { v_mul, 0, 0 },
    { 0, c, -s },
    { 0, s, c }
  }};
}

mat3 hsv_matrix(float h_angle, float s_mul, float v_mul) {
  return Yiq2Rgb * hsv_matrix_yiq(h_angle, s_mul, v_mul) * Rgb2Yiq;
}
```
4. The formula is wrong: should be `Yiq2Rgb * transform * Rgb2Yiq`, since the argument is on the right.